scipy/optimize/tests/test__shgo.py

```diff
@@ -610,32 +684,51 @@ def test_8_homology_group_diff(self):
 
     def test_9_cons_g(self):
         """Test single function constraint passing"""
-        SHGO(test3_1.f, test3_1.bounds, constraints=test3_1.cons[0])
+        SHGOc = SHGO(test3_1.f, test3_1.bounds, constraints=test3_1.cons[0])
 
     def test_10_finite_time(self):
```

Comment:
This test is purported to check that gh-13974 is resolved, but it is the same test as before. It appears to confirm only that `shgo` executes without error, but that was true before this PR, so it is not an adequate test of gh-13974.

Reply:
My understanding of gh-13974 is that it is that the `maxtime` option is not working properly because an extra iteration runs despite the allowed runtime being exceeded. This has been fixed here, I think the best way to test that is working as expected is as follows:

```python
    def test_10_finite_time(self):
        """Test single function constraint passing"""
        options = {'maxtime': 1e-15}
        def f(x):
            time.sleep(1e-14)
            return 0.0

        res = shgo(f, test1_1.bounds, n=1, iters=5,
                   options=options, sampling_method='sobol')

        # Assert that only 1 rather than 5 requested iterations ran:
        assert res.nit == 1
        assert res.tnev == 1
```
Logic:
* ```options = {'maxtime': 1e-15}``` limits shgo to `1e-15` s of runtime.
* The objective function takes a minimum of `1e-14` s of runtime.
* The requested number of is `iters=5`, therfore only a single iteration can be complete.
* Asserting that only a single iteration ran (`res.nit == 1`) before stopping the routine ensures that the option is working as expected. The assertion ` res.tnev` provides additional assurance since only one attemtped evaluation per iteration was specified.

I am not sure if there is a pitfall here, but in case you agree I have [updated the unittest](https://github.com/scipy/scipy/pull/16313/commits/2e53905b2ea16e54ebf08c6c6a4ad1ece092c9a0) to this code snippet.

Reply:
Yes, that looks about right. Since I don't really know what `n` does, the test would be a little more convincing to me if `n` were not set to `1`. To the uneducated reviewer, it looks like perhaps you're only getting one iteration because `n=1`, not because `maxtime` is set so low.

This will confirm that `maxtime` is doing what it is supposed to do, but several other options are mentioned in that issue. Sometimes it is hard to test that options are doing exactly what they are supposed to do, but I always try to write tests that show that options do _something_ that correlates with what they are supposed to do. 

If you have existing tests for all the supported options, please indicate which tests do that. If not, please add tests, or we can unlink gh-13974 and resolve it fully (with tests) in a separate PR. 

(Something else the issue mentions is that there is an interaction between the sample method and these options. If that's the case, we need to make sure that's all documented. If it already is, go ahead and point out that documentation in gh-13974.)

Reply:
I will replace it with:
```python
res = shgo(f, test1_1.bounds, iters=5,  options=options)
```
which stops at the first iteration, but produces more points. To briefly elaborate `maxtime` does not stop the algorithm mid iteration, only between each iteration. `n=1` was was used as this is the most likely intended use for maximum flexibility and the algorithm was supposed to sample `n*iters = 1*5 = 5` global points as opposed to `1` if it did not exceed 'maxtime'.

The two examples below illustrate the difference, first with `n=1`:
 ```python
options = {'maxtime': 3,
           'minimize_every_iter': False}

def f(x):
    time.sleep(1)
    return 0.0

res = shgo(f, test1_1.bounds, n=1, iters=5,
           options=options
           )
print(f'res.nfev = {res.nfev}')
print(f'res.nit = {res.nit}')
```
`res.nfev = 3`   (`n*iters = 1*3 = 3` function evaluations)
`res.nit = 3`  (3 seconds (+function overheard time) have passed which is > 3, therefore termination before iteration 4 starts)

Next to show `n=2` behaviour:

```python
options = {'maxtime':3,
           'minimize_every_iter': False}
def f(x):
    time.sleep(1)
    return 0.0

res = shgo(f, test1_1.bounds, n=2, iters=5,
           options=options
           )
print(f'res.nfev = {res.nfev}')
print(f'res.nit = {res.nit}')
```
outputs:
`res.nfev = 4`   (`n*iters = 2*2 = 4` function evaluations)
`res.nit = 2`  (4 seconds have passed which is > 3, therfore termination before iteration 4 starts)


Such tests would be more ideal, but add too much global time (heh) to the scipy test suite, and I am not sure how fine the time tolerance on overhead head is and this might cause failures in the future. Apologies for the amount of technical detail, I wanted to ensure there are no loopholes that I haven't thought of for this fix.

>If you have existing tests for all the supported options, please indicate which tests do that. 

I have the following unittests for three of the other options mentioned in that issue:
* 'maxiter': test_1_maxiter
* 'maxfev': test_3_1_no_min_pool_sobol
* 'f_min': test_4_1_known_f_min

'f_tol' is implicitly tested, but it really should be tested in `scipy.optimize.minimize` as it is only passed to the local routine and not used in the global routine.